### DISCUSSION TITLE
PrismaClientValidationError in tutorial

### DISCUSSION THREAD

**Discussion by gr-qft** [prisma/prisma#5513] [Q&A] [answered]
Hello,

I'm following along this tutorial 

https://vercel.com/guides/nextjs-prisma-postgres

At step 5, when handling the api call inside `pages/api/post/index.ts`, I've noticed that the line `author: { connect: { email: session?.user?.email } },` caused the following error: 

```js 
PrismaClientValidationError: 
Invalid `prisma.post.create()` invocation:

{
  data: {
    title: 'draft 1',
    content: 'draft 1 content ',
    author: {
    ~~~~~~
      connect: {
        email: null
      }
    }
  }
}
```

The full code handling the api call is: 

```js
export default async function handle(req, res) {
  const { title, content } = req.body

  const session = await getSession({ req })
  const result = await prisma.post.create({
    data: {
      title: title,
      content: content,
      author: { connect: { email: session?.user?.email } },
    },
  })
  res.json(result)
}
```

When I removed that line a new post was indeed created. 

Could somebody help me to understand what's going on here? 

Thanks!

**Comment by nikolasburk** [comment, marked as answer]
Hey @gr-qft! I just went through the tutorial until the step to create new drafts and I wasn't able to reproduce the error. Can you share a bit more context about your issue? Maybe even zip the project and upload it here so that I can take directly take a look to see what might be wrong? 🙂

**Reply by gr-qft** [reply]
@nikolasburk That's great, thanks, I really appreciate it. I'll push the code to a repo in the next couple of hours and I'll poke you then.

**Reply by gr-qft** [reply]
Hey @nikolasburk ! I've decided to upload the project here:

[code.zip](https://github.com/prisma/prisma/files/5948750/code.zip)

I'm seeing the following error when trying to create a new draft (step 5, after adding NextAuth): 

<img width="702" alt="Screen Shot 2021-02-08 at 10 42 29 PM" src="https://user-images.githubusercontent.com/69383359/107312783-35ae8800-6a5f-11eb-988b-07e5b8d254db.png">


I've also noticed the following:
- For some reason `.env` is created in the right place and the instruction `mv prisma/.env ./.env` in the tutorial does not apply. 
- There is a problem with NextAuth that requires  `next-auth@canary` (read https://github.com/prisma/prisma/discussions/5269)
- In the part that mentions Prisma adapter there's this line of code `  secret: process.env.SECRET`; however `SECRET` was never discussed. 
- I looked into the database and after adding NextAuth a logged-in user was indeed create in the database. But the email field is black. As far as I understand, a separate query to Github is required to get the email, but the tutorial never mentioned that. 

Thank you!

**Reply by nikolasburk** [reply]
Thank you so much for all that info! Weirdly enough, I still can't reproduce your error even with the zipped project you shared. For context: I did the just did the follwing:

1. downloaded your `code.zip` file
2. unzipped it
3. ran `yarn install`
4. added a `.env` file to the root with proper values for `DATABASE_URL`, `GITHUB_ID`, `GITHUB_SECRET` and `NEXTAUTH_URL`
5. ran `yarn prisma migrate dev --preview-feature` to create the DB and tables (could have also used `yarn prisma db push`, this shouldn't make a difference here)
6. started the app with `yarn dev`
7. was able to login and create a post 

So this is super weird! Maybe smth is wrong with the state of your DB somehow? Can you maybe retry running your project with a fresh database?

Also, thanks a lot for the additional hints!

> For some reason .env is created in the right place and the instruction mv prisma/.env ./.env in the tutorial does not apply.

Thanks for catching this! This behaviour had changed in Prisma since the tutorial was first written, will get that updated! 👍 

> There is a problem with NextAuth that requires next-auth@canary (read #5269)

Hmm, strange! Which version of the Prisma CLI are you using? Afaik this issue _should_ have been fixed in [this NextAuth release](https://twitter.com/nextauthjs/status/1356176212377423874?s=19). Please let me know if you still must use the `canary` version so I can follow up with my colleagues internally about this!

> In the part that mentions Prisma adapter there's this line of code secret: process.env.SECRET; however SECRET was never discussed.

Thanks for the hint, will try to improve this part in  the tutorial!

> I looked into the database and after adding NextAuth a logged-in user was indeed create in the database. But the email field is black. As far as I understand, a separate query to Github is required to get the email, but the tutorial never mentioned that.

Hmm, again something I can't reproduce unfortunately. After logging in with the project you provided as `code.zip`, the `email` field has been set 🤔 

![image](https://user-images.githubusercontent.com/4058327/107367346-78b04000-6adf-11eb-91e0-561270dd7e2f.png)

Hope that helps for now! I'm happy to dig deeper into this with you if you still encounter the same errors after you try running the project on a new DB!

**Reply by gr-qft** [reply]
Thanks a lot for the reply! I had some problem with Heroku database. Anyway, I created a new database and I'm not seeing the original error anymore, but the email column is not showing. 

<img width="1043" alt="Screen Shot 2021-02-11 at 8 28 50 AM" src="https://user-images.githubusercontent.com/69383359/107642690-34d64b80-6c43-11eb-956b-b75464ea0cda.png">

What could be going on here? 

I'll look into whether or not if I still need to use `canary` NextAuth version later today and I'll let you know then. Also, once this email is issue is resolved, I'll work through the rest of the tutorial and let you know if everything is good. 

Thanks.

**Reply by gr-qft** [reply]
@nikolasburk Sorry I've not had a chance to re-visit this problem. I'll be able to come back to it tomorrow, and I'll poke you then. Hope you don't mind.

Thanks.

**Reply by gr-qft** [reply]
@nikolasburk It's correct that `next-auth@canary` is not necessary anymore. However, I'm still not able to see users' emails after logging in with Github. I would like to get this problem out of the way first. What are you suggesting?

Thanks!

**Reply by nikolasburk** [reply]
> However, I'm still not able to see users' emails after logging in with Github. I would like to get this problem out of the way first. What are you suggesting?

Hmm good question, I'm not sure to be honest. I think this is rather an issue with NextAuth rather than Prisma, so it's probably better to ask someone from their team to get a proper response for this 😄 

My hunch is that the email is only stored when it's part of a public profile. Since I have my email public on GitHub but you don't, this _might_ be the reason...

![image](https://user-images.githubusercontent.com/4058327/108326307-62367280-71ca-11eb-8b8c-9648932f0b04.png)

![image](https://user-images.githubusercontent.com/4058327/108326340-69f61700-71ca-11eb-82c3-bc78605d3e2d.png)

**Reply by gr-qft** [reply]
Thank you for the hint! You're right! I made my email public and the email address showed up on the app page. I think the tutorial should be updated regarding this. By the way, if an email is set private it's still possible to get it with an extra fetch. 

I'll proceed to work through the rest of the tutorial.

**Reply by gr-qft** [reply]
@nikolasburk I confirm that everything in the tutorial works. I think the key thing to note is that the tutorial assumes github users' emails will be visible publicly, otherwise it seems things break down from step 5.

**Reply by OmanCoding** [reply]
I actually encountered the same problem, I could not use create with connect for a field due to Unknown arg

However, after creating it, I used .update with connect and it works.

For some reason, it is not seen on create()

**Comment by gr-qft** [comment]
@nikolasburk Just another question: inside `lib/prisma.ts` why do we check if it's production or not, is it for caching? 

By the way, I got Typescript warning that `prisma` does not exist on `Global`. I think to remove the warning, it could be written as 

```js
import { PrismaClient } from '@prisma/client'

let prisma: PrismaClient

if (process.env.NODE_ENV === 'production') {
    prisma = new PrismaClient()
} else {
    if (!global['prisma']) {
        global['prisma'] = new PrismaClient()
    }
    prisma = global['prisma']
}

export default prisma
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
